### ISSUE TITLE
dotnet build/publish uses the implicit SDK RID for RID-specific apps by default

### ISSUE THREAD

**Issue by richlander** [dotnet/sdk#23539]
# dotnet build/publish uses the implicit SDK RID for RID-specific apps by default

**Proposal:** .NET SDK uses the implicit Runtime Identifier (RID) in each case when a RID is required but isn't specified by the user. It should affect (at least) the following `dotnet` commands:

* `build`
* `publish`
* `run`

We should validate where `--self-contained` is supported/valid. Wherever it is valid, this behavior change should be made.

**Example:** The following example (console app template) should not fail, but succeed (using the implicit SDK RID). This failure isn't useful (it doesn't protect the user from anything scary).

```bash
PS C:\Users\rich\app3> dotnet build --self-contained
Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\6.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(150,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [C:\Users\rich\app3\app3.csproj]

Build FAILED.
```

Note: Part of this feature was implemented in https://github.com/dotnet/sdk/pull/22314.

**Comment by imba-tjd** [comment]
What about

```cmd
dotnet publish -p:PublishReadyToRun=true
```

and

```cmd
dotnet publish --self-contained=false -p:PublishReadyToRun=true
```

Currently they must specify RID.

**Comment by Naine** [comment]
`PublishSingleFile` too. The current proposal seems to cover them:

> in each case when a RID is required but isn't specified by the user

**Comment by tmds** [comment]
This should use the _portable rid_, and not the rid the SDK/runtime was built for.

.NET 6 SDK added an option for this called `--use-current-runtime` to `build`/`restore`/`publish` but not `run`.

Adding the rid affects the restore behavior and causes more dependencies to be downloaded. Probably this is not desired.

**Comment by richlander** [comment]
Just saw this. This doesn't ring true for me. There is no such option today, opt-in or otherwise, other than perhaps `-r any`. Also, what would portable buy you since you have a rid-specific executable and no other way to launch the app than that executable?

**Comment by tmds** [comment]
> There is no such option today

There is `--use-current-runtime`.

> other than perhaps -r any. Also, what would portable buy you since you have a rid-specific executable and no other way to launch the app than that executable?

By portable, I don't mean `any`. I mean the associated portable rid of the platform. For example, on a source-built SDK which has the non-portable rid `fedora.35-x64`, the associated portable rid is `linux-x64`.

The SDK already stores this rid in the `Microsoft.NETCoreSdk.BundledVersions.props` file as `NETCoreSdkPortableRuntimeIdentifier`.

> in each case when a RID is required

https://github.com/dotnet/sdk/issues/23545 is proposing: _dotnet build/publish does not produce an exe/apphost for portable apps by default_.

So, if I understand correctly:
When these commands are used without any arguments, they will no longer require a RID.
And, this issue applies when the user adds an argument that requires a RID, like: `--self-contained`.

**Comment by baronfel** [comment]
Yeah, `--use-current-runtime` eventually resolves to the portable RID:
* it sets the `UseCurrentRuntimeIdentifier=true` msbuild property,
* in the runtime inference targets, the presence of that property sets the `RuntimeIdentifier` to `NETCoreSdkPortableRuntimeIdentifier`
* `NETCoreSdkPortableRuntimeIdentifier` is set in dotnet/installer to `PortableProductMonikerRid`
* `PortableProductMonikerRid`  is also set in dotnet/installer to `$(HostOSName)-$(Architecture)`, where `HostOSName` is one of `win`, `osx`, `freebsd`, or `linux`

So yes, I believe they should be identical.

**Comment by richlander** [comment]
Ya. I was mentioning this to @agocke. We should do all those things. Add `PublishAot` to the list.

**Comment by agocke** [comment]
I don't think anything except SelfContained probably needs anything.

If you have PublishAot in your project file, you don't want `dotnet build` to suddenly start producing self-contained builds. And the rest of them aren't even self-contained specific during publish, so I don't think there's anything to do there.

**Comment by richlander** [comment]
Oh my, did I just realize that we don't have a `PublishSelfContained` property? If not, that seems like an odd hole! Can we add that?

**Comment by agocke** [comment]
Is the goal to allow Publish to behave differently from build?

**Comment by richlander** [comment]
No. This is agnostic to build/publish. It's whatever argument requires a RID can use the implicit one. Most of the cases for that are in `publish` but not exclusively. Even if they were only on `publish`, the answer would still be no. This isn't about `build` vs `publish`.

**Comment by richlander** [comment]
I wrote a mini-spec for the `Publish*` properties at https://github.com/dotnet/sdk/issues/26028

**Comment by richlander** [comment]
I wrote a mini-spec on FDD + (implicit) RID-specific apps at https://github.com/dotnet/sdk/issues/26031

**Comment by nagilson** [comment]
This should be resolved in the release/7.0.1xx branch now. I'd imagine it will flow into rc-2 within a week or two. Please let me know if you encounter any issues. Elsewise we can mark this as resolved! Thanks for working with us to improve the end user experience
